pennylane/math/tensorbox.py

```diff
@@ -12,10 +12,34 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """This module contains the TensorBox abstract base class."""
-# pylint: disable=import-outside-toplevel
+# pylint: disable=import-outside-toplevel,too-many-public-methods
 import abc
+import functools
+import numbers
+import sys
 
-import numpy as np
+
+def wrap_output(func):
+    """Decorator to automate the wrapping of TensorBox method outputs.
+
+    When applied to a TensorBox method, it inserts an additional argument
+    into the signature, ``wrap_output``. By default, this is True, causing
+    the output of the method to be wrapped as a TensorBox; specifying
+    ``wrap_output=False`` when calling the method results in the
+    underlying tensor itself being returned.
+    """
+
+    @functools.wraps(func)
+    def _wrapper(*args, **kwargs):
+        wrap = kwargs.pop("wrap_output", True)
+
+        if wrap:
+            cls = vars(sys.modules[func.__module__])[func.__qualname__.split(".")[0]]
+            return cls(func(*args, **kwargs))
+
+        return func(*args, **kwargs)
+
+    return _wrapper

```

Comment:
This should be quite a fast decorator, since it's not doing anything really 'clever'. In the past, we saw slowdowns with decorators that explicitly used `inspect.signature` to perform logic based on the function signature.

Reply:
I see, got it!